src/framework/FormatChecker.ts

```diff
@@ -0,0 +1,102 @@
+import { Base, BaseConfig } from './Base';
+import { Bot } from './Bot';
+import { findTextChannelByName } from './helpers';
+import { Message } from 'discord.js';
+import uuid from '@lukeed/uuid';
+import { BaseLogger } from 'pino';
+
+type FunctionChecker = (message: string, logger?: BaseLogger) => boolean;
+
+export interface FormatCheckerConfig extends BaseConfig {
+  channelName: string;
+  checker: RegExp | FunctionChecker;
+  examples?: string[];
+}
+
+export class FormatChecker extends Base {
+  private readonly channelName: string;
+  private readonly checker: RegExp | FunctionChecker;
+  private readonly examples?: string[];
+
+  private bot: Bot | undefined;
+
+  public constructor(config: FormatCheckerConfig) {
+    super(config);
+    this.channelName = config.channelName;
+    this.checker = config.checker;
+    this.examples = config.examples;
+
+    this._messageHandler = this._messageHandler.bind(this);
+  }
+
+  async _messageHandler(message: Message): Promise<void> {
+    if (this.bot === undefined) return;
+    const client = this.bot.client;
+    if (
+      message.channel.id !== findTextChannelByName(client, this.channelName)?.id
+    )
+      return;
+
+    const logger = this.bot.logger.child({
+      id: uuid(),
+      type: 'FormatChecker',
+      name: this.name,
+    });
+
+    if (this.checker instanceof RegExp) {
+      if (this.checker.test(message.cleanContent) === true) return;
+    } else if (this.checker instanceof Function) {
+      if (this.checker(message.cleanContent, logger)) return;
+    } else {
+      logger.error(`invalid checker for ${this.name}`);
```

Comment:
can we throw this during construction instead of here? the checker cannot change at runtime so early validation is better for the developer